### PR TITLE
[BugFix] Exactly-once semantic enhancement

### PR DESCRIPTION
**Pull request by xlfjcg**

### CODE REVIEW

**Comment by banmoy** [comment]
There are other two problems
1. Current transaction should not be committed in `StarRocksDynamicSinkFunctionV2#finish`, and just `flush` is enough. You can refer to the comment of [StreamOperator#finish](https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperator.java#L83)
2. In `StarRocksDynamicSinkFunctionV2#close`, those transactions not in a checkpoint should be aborted.

**Comment by xlfjcg** [comment]
1. should test snapshotState or notifyCheckpointComplete will trigger after finish. 
2. At least once may be useful

File: src/main/java/com/starrocks/connector/flink/table/sink/StarRocksDynamicSinkFunctionV2.java [hunk 1]
```diff
@@ -195,36 +180,67 @@ public void snapshotState(FunctionSnapshotContext functionSnapshotContext) throw
...
+            snapshotStates.add(snapshotMap);
         } else {
             throw new RuntimeException("Snapshot state failed by prepare");
         }
     }
 
     @Override
     public void initializeState(FunctionInitializationContext functionInitializationContext) throws Exception {
-        ListStateDescriptor<Tuple2<Long, StreamLoadSnapshot>> descriptor =
+        ListStateDescriptor<Map<Long, List<StreamLoadSnapshot>>> descriptor =
```
Comment (banmoy) [inline src/main/java/com/starrocks/connector/flink/table/sink/StarRocksDynamicSinkFunctionV2.java:194]: Can we encapsulate the state in a class like `Snapshot`, and define a serializer supports state compatibility (for example, write a version before serializer the snapshot data)? We can't always rely on state name to ensure compatibility (>﹏<).

File: pom.xml [hunk 2]
```diff
@@ -43,8 +43,11 @@ limitations under the License.
     <modelVersion>4.0.0</modelVersion>
     <groupId>com.starrocks</groupId>
     <artifactId>flink-connector-starrocks</artifactId>
-    <version>1.2.3_flink-1.15</version>
+    <version>${srfc.version}_flink-1.15</version>
```
Comment (chaoyli) [inline pom.xml:46]: You can add a more clear commit message about the problem, solution and results

File: src/main/java/com/starrocks/connector/flink/table/sink/StarRocksDynamicSinkFunctionV2.java [hunk 3]
```diff
@@ -195,36 +181,65 @@ public void snapshotState(FunctionSnapshotContext functionSnapshotContext) throw
...
     @Override
     public void notifyCheckpointComplete(long checkpointId) throws Exception {
-        StreamLoadSnapshot snapshot = snapshotMap.get(checkpointId);
-        if (snapshot == null) {
-            log.warn("Unknown checkpoint id : {}", checkpointId);
-            return;
+
+        boolean succeed = true;
+
+        List<Long> commitCheckpointIds = snapshotMap.keySet().stream()
```
Comment (chaoyli) [inline src/main/java/com/starrocks/connector/flink/table/sink/StarRocksDynamicSinkFunctionV2.java:222]: Could the commitCheckpointIds be a empty list?

File: pom.xml (hunk 2 above)
Comment (xlfjcg) [reply pom.xml:46]: flink state should contains srfc.version, flag the struct version

File: src/main/java/com/starrocks/connector/flink/table/sink/StarRocksDynamicSinkFunctionV2.java (hunk 3 above)
Comment (xlfjcg) [reply src/main/java/com/starrocks/connector/flink/table/sink/StarRocksDynamicSinkFunctionV2.java:222]: empty is possible

**Comment by banmoy** [comment]
LGTM

**Review by chaoyli — APPROVED** [review]

File: src/main/java/com/starrocks/connector/flink/table/sink/StarRocksDynamicSinkFunctionV2.java [hunk 4]
```diff
@@ -159,9 +161,11 @@ public void open(Configuration parameters) throws Exception {
             rowTransformer.setRuntimeContext(getRuntimeContext());
         }
 
-        for (Tuple2<Long, StreamLoadSnapshot> tuple : snapshotStates.get()) {
-            if (!sinkManager.commit(tuple.f1)) {
-                snapshotMap.put(tuple.f0, tuple.f1);
+        synchronized (snapshotMap) {
+            for (Tuple2<Long, StreamLoadSnapshot> tuple : snapshotStates.get()) {
```
[the remark this answers is not in the document]
Comment (hackeryang) [reply src/main/java/com/starrocks/connector/flink/table/sink/StarRocksDynamicSinkFunctionV2.java]: Hello @banmoy , I would like to ask why we didn't implement the interface `TwoPhaseCommitSinkFunction` for `exactly-once`, but encapsulated another interface `StarRocksDynamicSinkFunctionBase` with `StarRocksDynamicSinkFunctionV2`. Are there other considerations, such as performance issues?

**Pull request merged**